lib/index.d.ts

```diff
@@ -0,0 +1,528 @@
+    revocation_endpoint: string
+    end_session_endpoint: string
+    registration_endpoint: string
+    token_endpoint_auth_methods_supported: string
+    token_endpoint_auth_signing_alg_values_supported: string
+    introspection_endpoint_auth_methods_supported: string
+    introspection_endpoint_auth_signing_alg_values_supported: string
+    revocation_endpoint_auth_methods_supported: string
+    revocation_endpoint_auth_signing_alg_values_supported: string
+    request_object_signing_alg_values_supported: string
+    mtls_endpoint_aliases: IMetadataMtlsEndpointAliases
+
+    [key: string]: any
+  }
+
+  export interface IMetadataMtlsEndpointAliases {
+    token_endpoint: string
+    userinfo_endpoint: string
+    revocation_endpoint: string
+    introspection_endpoint: string
+  }
+
+  /**
+   * Encapsulates a discovered or instantiated OpenID Connect Issuer (Issuer), Identity Provider (IdP),
+   * Authorization Server (AS) and its metadata.
+   */
+  export interface IIssuer {
+    /**
+     * Returns the <Client> class tied to this issuer.
+     */
+    Client: { new (metadata: IClientMetadata, jwks?: object) }
+
+    /**
+     * Returns metadata from the issuer's discovery document.
+     */
+    metadata?: IIssuerMetadata
+
+    /**
+     * Returns the issuer's jwks_uri keys as a @panva/jose parsed JWKS.Keystore.
+     * @param forceReload
+     */
+    keystore (forceReload?: boolean): Promise<JWKS.KeyStore>
+
+    [key: string]: any
+  }
+
+  export class Issuer implements IIssuer {
+    constructor (metadata: IIssuerMetadata)
+    Client: typeof Client
+    metadata: IIssuerMetadata
+    keystore (forceReload?: boolean): Promise<JWKS.KeyStore>
+
+    /**
+     * Loads OpenID Connect 1.0 and/or OAuth 2.0 Authorization Server Metadata documents.
+     * When the issuer argument contains '.well-known' only that document is loaded, otherwise
+     * performs both openid-configuration and oauth-authorization-server requests.
+     * @param issuer Issuer Identifier or metadata URL
+     */
+    static discover (issuer: string): Promise<Issuer>
+
+    /**
+     * Performs OpenID Provider Issuer Discovery based on End-User input.
+     * @param input EMAIL, URL, Hostname and Port, acct or syntax input
+     */
+    static webfinger (input: string): Promise<Issuer>
+  }
+
+  export interface ITokenSetParams {
+    /**
+     * The raw access token in JWT format
+     */
+    access_token?: string
+    /**
+     * Usually "Bearer"
+     */
+    token_type?: string
+    /**
+     * The raw id token in JWT format
+     */
+    id_token?: string
+    /**
+     * Refresh token, opaque string value
+     */
+    refresh_token?: string
+    /**
+     * space-separated scope(s) used for the authentication request
+     */
+    scope?: string
+    expires_in?: number
+    expires_at?: number
+    /**
+     * State value passed in the authentication request
+     */
+    session_state?: string | object
+
+    [key: string]: any
```

Comment:
`unknown` again

Reply:
on second thought, should we allow this to be extensible at all? I can't think of an instance where a `TokenSet` will have additional properties, @panva may be better placed to comment

Reply:
Anything returned by the token response is available actually. So yeah. But as you suggest above, unknown as a type.